### PR TITLE
Add Custom Config merger for Kafka - 3.4.0 

### PR DESCRIPTION
**Pull request by obaydullahmhs**
Signed-off-by: obaydullahmhs <obaydullah@appscode.com>

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -44,21 +130,19 @@ delete_cluster_metadata() {
...
-      echo "$CLUSTER_ID" > "$log_dirs"/cluster_id
-  else
-      CLUSTER_ID=$(cat "$log_dirs"/cluster_id)
+  if [[ -d "$metadata_log_dir/__cluster_metadata-0" ]]; then
+     rm -rf $metadata_log_dir/meta.properties
   fi
-
+  if [ -e "$metadata_log_dir/__cluster_metadata-0/quorum-state" ] ; then
+     rm -rf "$metadata_log_dir/__cluster_metadata-0/quorum-state"
+  fi
```
Comment (raihankhan) [inline entrypoint.sh]: Add comment > why we should delete cluster quorum state

File: entrypoint.sh [hunk 2]
```diff
@@ -8,8 +8,94 @@ temp_clientauth_config="/opt/kafka/config/temp-config/clientauth.properties"
...
+custom_config_dir="/opt/kafka/config/custom-config"
+
+merge_custom_config() {
+
+    # Merge the custom config with the default config
+    echo "Merging custom config with default config"
+
+    MASTER_FILE="/opt/kafka/config/custom-config/config.properties"
+    SLAVE_FILE=$operator_config
+    OUTPUT_FILE="/opt/kafka/config/kafkaconfig/config.properties.merged"
```
Comment (raihankhan) [inline entrypoint.sh]: Convert all env var to small case

File: entrypoint.sh [hunk 3]
```diff
@@ -8,8 +8,94 @@ temp_clientauth_config="/opt/kafka/config/temp-config/clientauth.properties"
...
+
+    if [[ ! -d "$custom_config_dir" ]]; then
+      echo "No custom configuration found"
```
Comment (raihankhan) [inline entrypoint.sh]: Remove this log

File: entrypoint.sh [hunk 4]
```diff
@@ -8,8 +8,94 @@ temp_clientauth_config="/opt/kafka/config/temp-config/clientauth.properties"
...
+      return
+    fi
+    #ignore the following properties while merging
+    arr=("process.roles", "cluster.id", "node.id", "controller.quorum.voters",
```
Comment (raihankhan) [inline entrypoint.sh]: Try the Array without commas

File: entrypoint.sh [hunk 5]
```diff
@@ -8,8 +8,94 @@ temp_clientauth_config="/opt/kafka/config/temp-config/clientauth.properties"
...
+    "control.plane.listener.name", "listeners", "advertised.listeners")
+    # Delete the output file if exists
```
Comment (raihankhan) [inline entrypoint.sh]: Mention previous

File: entrypoint.sh [hunk 6]
```diff
@@ -8,8 +8,94 @@ temp_clientauth_config="/opt/kafka/config/temp-config/clientauth.properties"
...
+    if [ -e $OUTPUT_FILE ] ; then
+        rm -rf "$OUTPUT_FILE"
+    fi
```
Comment (raihankhan) [inline entrypoint.sh]: Remove this part if unnecessary

File: Dockerfile [hunk 7]
```diff
@@ -17,7 +17,8 @@ RUN apt-get update \
  && tar xfz /tmp/kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz -C /opt \
  && rm /tmp/kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz \
  && mv /opt/kafka_${SCALA_VERSION}-${KAFKA_VERSION} ${HOME} \
- && rm -rf /opt/kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz
+ && rm -rf /opt/kafka_${SCALA_VERSION}-${KAFKA_VERSION}.tgz \
+ && apt-get install kafkacat -y
```
Comment (raihankhan) [inline Dockerfile:22]: Add ref link for kafkacat

**Review by raihankhan — CHANGES_REQUESTED** [review]

**Review by raihankhan — APPROVED** [review]
LGTM!

**Review by heheh13 — APPROVED** [review]

**Pull request merged**